### PR TITLE
MAJ-2 Delayed tasks

### PR DESCRIPTION
**Pull request by caiok**

### CODE REVIEW

File: majordodo-core/src/main/java/majordodo/task/StatusEdit.java [hunk 1]
```diff
@@ -316,6 +322,7 @@ public static final StatusEdit WORKER_DIED(String workerId, long timestamp) {
                     doo.writeUTF(taskType);
                     doo.writeInt(maxattempts);
                     doo.writeInt(attempt);
+                    doo.writeLong(requestedStartTime);
```
Comment (eolivelli) [inline majordodo-core/src/main/java/majordodo/task/StatusEdit.java]: this is going to break compatibility on the format of the log:
- you will not be able to upgrade a broker
- followers will not be able to follow a broker of an previous version

it is important that new followers can follow "old leaders", in a rolling restart you are going to restart followers if possible and than the leader

File: majordodo-core/src/main/java/majordodo/task/Task.java [hunk 2]
```diff
@@ -234,10 +247,42 @@ Task cloneForSnapshot() {
...
         return copy;
     }
 
+    @Override
+    public long getDelay(TimeUnit unit) {
+        long delay = requestedStartTime - System.currentTimeMillis();
+        if (delay < 0) {
+            delay = 0;
+        }
+        switch (unit) {
```
Comment (eolivelli) [inline majordodo-core/src/main/java/majordodo/task/Task.java]: are you sure ?
I think usually this kind of code does not enumerate all possible cases. IMHO the is a bug here

this is a better code:
`return unit.convert(requestedStartTime - System.currentTimeMillis(), TimeUnit.MILLISECONDS);`

**Comment by eolivelli** [comment]
findbugs is reporting errors:
```
[INFO] --- findbugs-maven-plugin:3.0.4:check (default-cli) @ majordodo-core ---
[INFO] BugInstance size is 5
[INFO] Error size is 0
[INFO] Total bugs: 5
[INFO] Nullcheck of taskId at line 756 of value previously dereferenced in majordodo.task.Broker.addTasks(List) [majordodo.task.Broker, majordodo.task.Broker] At Broker.java:[line 744]Redundant null check at Broker.java:[line 756] RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE
[INFO] Integral division result cast to double or float in majordodo.task.BrokerStatusMonitor$Monitor.run() [majordodo.task.BrokerStatusMonitor$Monitor] At BrokerStatusMonitor.java:[line 96] ICAST_IDIV_CAST_TO_DOUBLE
[INFO] Dead store to version in majordodo.task.StatusEdit.read(byte[]) [majordodo.task.StatusEdit] At StatusEdit.java:[line 662] DLS_DEAD_LOCAL_STORE
[INFO] majordodo.task.StatusEdit defines equals and uses Object.hashCode() [majordodo.task.StatusEdit] At StatusEdit.java:[lines 186-262] HE_EQUALS_USE_HASHCODE
[INFO] majordodo.task.Task defines compareTo(Object) and uses Object.equals() [majordodo.task.Task] At Task.java:[line 32] EQ_COMPARETO_USE_OBJECT_EQUALS
[INFO] 

```

**Comment by caiok** [comment]
Already integrated in MD 0.11.0

**Pull request closed without merging**